x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,542 @@
+import org.apache.lucene.util.automaton.Automaton;
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.lucene.BytesRefs;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.plain.DocValuesIndexFieldData;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.Mapper;
+import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.ParseContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.index.mapper.TypeParsers.parseField;
+
+/**
+ * A {@link FieldMapper} for indexing fields with ngrams for efficient wildcard matching
+ */
+public class WildcardFieldMapper extends FieldMapper {
+
+    public static final String CONTENT_TYPE = "wildcard";
+    public static short MAX_NUM_CHARS_COUNT = 6; //maximum allowed number of characters per ngram
+
+    public static class Defaults {
+        public static final MappedFieldType FIELD_TYPE = new WildcardFieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.setStoreTermVectorOffsets(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.freeze();
+        }
+        public static final int IGNORE_ABOVE = Integer.MAX_VALUE;        
+    }
+
+    public static class Builder extends FieldMapper.Builder<Builder, WildcardFieldMapper> {
+        private int numChars = 3;
+        protected int ignoreAbove = Defaults.IGNORE_ABOVE;
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        @Override
+        public Builder docValues(boolean docValues) {
+            if (docValues == false) {
+                throw new MapperParsingException("The field [" + name +
+                        "] cannot have doc values = false");                
+            }
+            return this;
+        }
+        
+        @Override
+        public Builder index(boolean index) {
+            if (index == false) {
+                throw new MapperParsingException("The field [" + name +
+                        "] cannot have index = false");                
+            }
+            return this;
+        }
+
+        public Builder numChars(int numChars) {
```

Comment:
nit: I wonder if this is needed ? Trade-offs are different when using bigger grams and I like the fact that this is an implementation detail that the user should not care about.

Reply:
The default of 3grams seems like something that doesn't need to be tinkered with. 5grams was painfully slow to index. I'd be happy to remove configurability.

Re the need for TaperedNGramTokenFilter - it helps with searches like `*foobar*.c` where the end of the pattern is < 3.

Reply:
> Re the need for TaperedNGramTokenFilter - it helps with searches like *foobar*.c where the end of the pattern is < 3.

That's the part that I am not sure of since we have to add the "end of stream" character so `*.c` is in fact two characters. Prefix queries of 2 characters on a field indexed with a max of 3 shouldn't be too costly ?